### PR TITLE
Remove AppSec::instrumentation::Gateway#pushed?

### PR DESCRIPTION
**Pull request by y9v**
**What does this PR do?**
Moves identity event tracking from the global `Gateway` singleton to the per-request AppSec Context. Removes `Gateway#pushed?` and `@pushed_events` entirely.

The fix adds an identity_event_collected flag to the per-request `Context#state` hash, set by the identity watchers (`watch_user_id`, `watch_user_login`), and read by `watch_request_finish`. This correctly scopes header collection to the request that triggered the identity event.

**Motivation:**
`Gateway#pushed?` tracked events on a process-lifetime singleton that was never reset between requests. Once any request pushed an identity event (e.g. a Devise login), `pushed?` returned `true` for all subsequent requests forever. This caused watch_request_finish to collect identity headers on every request after the first login, rather than only on requests where an identity event actually occurred.

**Change log entry**
None.

**Additional Notes:**
Related to #5466

**How to test the change?**
CI and manual testing

### CODE REVIEW

File: lib/datadog/appsec/context.rb [hunk 1]
```diff
@@ -62,7 +62,8 @@ def initialize(trace, span, waf_runner)
         @state = {
           events: [],
           interrupted: false,
-          downstream_body_analyzed_count: 0
+          downstream_body_analyzed_count: 0,
+          identity_event_collected: false
```
Comment (Strech) [inline lib/datadog/appsec/context.rb]: I think it should work without the key too, right? I have a feeling that we might miss one spec - when nothing present in the state.

File: spec/datadog/appsec/integration/contrib/devise/signin_multi_user_tracking_spec.rb [hunk 2]
```diff
@@ -223,8 +223,6 @@ def new
         '_dd.appsec.usr.login' => 'john.doe@example.com',
         '_dd.appsec.usr.id' => 'user:1'
       )
-
-      expect(gateway.pushed?('appsec.events.user_lifecycle')).to be true
```
Comment (Strech) [inline spec/datadog/appsec/integration/contrib/devise/signin_multi_user_tracking_spec.rb:227]: As we remove that check, to avoid similar mistake, does it make sense to add tests to verify

1. It's working when state is empty (no key)
2. It's working when state is filled (false key)
3. It's not leaking between the requests (current bug)

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb [hunk 3]
```diff
@@ -116,7 +116,7 @@ def watch_request_finish(gateway = Instrumentation.gateway)
                 gateway.watch('rack.request.finish') do |stack, gateway_request|
                   context = gateway_request.env[AppSec::Ext::CONTEXT_KEY]
 
-                  if context.span.nil? || !gateway.pushed?('appsec.events.user_lifecycle')
+                  if context.span.nil? || !context.state[:identity_event_collected]
```
Comment (Strech) [inline lib/datadog/appsec/contrib/rack/gateway/watcher.rb]: I think this name is misleading, before the state use it was name reflecting that an event of identity lifecycle (login, signup or auth request) had happen and at this point we know that it was.

With wording "event collected", I'm assuming something already processed that event, but in fact this is quite opposite - as it is "event produced" or "event happened" kind-a thing.

Renaming suggestions (by AI):

 Five suggestions, production-oriented:

  1. `identity_event_triggered` — direct, says the event fired, no ambiguity about direction
  2. `identity_event_occurred` — neutral/factual, just states it happened
  3. `user_lifecycle_event_fired` — closest to the original semantic, keeps "lifecycle"
  4. `identity_event_detected` — implies the system noticed it happened, slight observer framing but still correct direction
  5. `has_identity_event` — predicate style, reads naturally as a Hash key (`state[:has_identity_event]`), very Ruby-idiomatic

_P.S my personal vote goes to 4 or 5_

File: spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb [hunk 4]
```diff
@@ -112,4 +113,23 @@
...
+      clear_traces!
+
+      get('/without-identity-set-user', {}, headers)
+    end
+
+    it 'does not collect identity related request headers for the second request' do
+      expect(response).to be_ok
+
+      expect(http_service_entry_span.tags).not_to have_key('http.request.headers.cf-connecting-ipv6')
+    end
```
Comment (Strech) [inline spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb]: Just to be sure, let's do positive testing here too.

1. get + header
2. clear
3. get + no header

Just to cover everything

File: lib/datadog/appsec/contrib/rack/gateway/watcher.rb (hunk 3 above)
Comment (y9v) [reply lib/datadog/appsec/contrib/rack/gateway/watcher.rb]: good point, renamed to `has_identity_event` in [09c7995](https://github.com/DataDog/dd-trace-rb/pull/5487/commits/09c7995e6bca6609c2f90aded9914f2ef5463b9f)

File: spec/datadog/appsec/integration/contrib/devise/signin_multi_user_tracking_spec.rb (hunk 2 above)
Comment (y9v) [reply spec/datadog/appsec/integration/contrib/devise/signin_multi_user_tracking_spec.rb:227]: We have tests for this behaviour in `request_headers_collection_for_identity_spec.rb` - I also added a test to detect if the state is leaking from request to request.

Do you want me to add some unit tests for this? I think having a test for empty state and where the key is set to false is good, but it's more a unit test, not an integration test, since we never set the key to `false` in our codebase.

Comment (Strech) [reply spec/datadog/appsec/integration/contrib/devise/signin_multi_user_tracking_spec.rb:227]: Fine with `request_headers_collection_for_identity_spec.rb` to cover it 👍🏼

File: spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb (hunk 4 above)
Comment (y9v) [reply spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb]: done, added a positive test in [59fbd3a](https://github.com/DataDog/dd-trace-rb/pull/5487/commits/59fbd3ad425e529e92e75a0331a92f7dfb1426f5)

File: spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb [hunk 5]
```diff
@@ -111,5 +114,37 @@
...
+      )
+    end
+
+    it 'does not leak identity headers into the second request' do
+      clear_traces!
+      get('/without-identity-set-user', {}, {'HTTP_CF_CONNECTING_IPV6' => '2001:db8:3333:4444:5555:6666:1.2.3.4'})
+
+      expect(last_response).to be_ok
+      expect(http_service_entry_span.tags).not_to have_key('http.request.headers.cf-connecting-ipv6')
+    end
```
Comment (Strech) [inline spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb]:
```suggestion
    let(:headers) do
      {'HTTP_CF_CONNECTING_IPV6' => '2001:db8:3333:4444:5555:6666:1.2.3.4'}
    end

    it 'does not leak identity headers into the second request' do
      get('/with-identity-set-user', {}, headers)
      
      expect(last_response).to be_ok
      expect(http_service_entry_span.tags).to include(
        'http.request.headers.cf-connecting-ipv6' => '2001:db8:3333:4444:5555:6666:1.2.3.4'
      )
    
      clear_traces!

      get('/without-identity-set-user', {}, headers)

      expect(last_response).to be_ok
      expect(http_service_entry_span.tags).not_to have_key('http.request.headers.cf-connecting-ipv6')
    end
```

Comment (y9v) [reply spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb]: :+1: nice, thanks!

Comment (y9v) [reply spec/datadog/appsec/integration/contrib/rack/request_headers_collection_for_identity_spec.rb]: [4489ca1](https://github.com/DataDog/dd-trace-rb/pull/5487/commits/4489ca17f823c94cb6afe618374f5906e622200d)
